arcade/sprite_list/spatial_hash.py

```diff
@@ -1,13 +1,54 @@
+from abc import abstractmethod
 from math import trunc
 from typing import Generic
 
-from arcade.sprite import SpriteType
+from arcade.sprite import SpriteType, SpriteType_co
 from arcade.sprite.base import BasicSprite
 from arcade.types import IPoint, Point
 from arcade.types.rect import Rect
 
 
-class SpatialHash(Generic[SpriteType]):
+class ReadOnlySpatialHash(Generic[SpriteType_co]):
+    """
+    Read-only view of a SpatialHash.
+
+    This is useful when the SpriteType is in covariant position.
+
+    See SpatialHash for more details.
+    """
+
+    @abstractmethod
+    def get_sprites_near_sprite(self, sprite: BasicSprite) -> set[SpriteType_co]:
+        """
+        Get all the sprites that are in the same buckets as the given sprite.
+
+        Args:
+            sprite: The sprite to check
+        """
+        ...
+
+    @abstractmethod
+    def get_sprites_near_point(self, point: Point) -> set[SpriteType_co]:
+        """
+        Return sprites in the same bucket as the given point.
+
+        Args:
+            point: The point to check
+        """
+        ...
+
+    @abstractmethod
+    def get_sprites_near_rect(self, rect: Rect) -> set[SpriteType_co]:
+        """
+        Return sprites in the same buckets as the given rectangle.
+
+        Args:
+            rect: The rectangle to check (left, right, bottom, top)
```

Comment:
If we're adding this, we may as well correct it.
```suggestion
    .. tip:: Use :py:mod:`arcade.types.rect`'s helper functions to create rectangle objects!

    Args:
        rect: The rectangle to check as a :py:class:`~arcade.types.rect.Rect` object.
```